src/common/io.ts

```diff
@@ -1000,6 +1004,72 @@ export class ARIOReadable implements AoARIORead, ArNSNameResolver {
       type: nameData.type,
     };
   }
+
+  /**
+   * Get all ARNS names associated with an address using the provided ANT registry address.
+   *
+   * By default it will use the mainnet ANT registry address.
+   *
+   * @param {Object} params - The parameters for fetching ARNS names
+   * @param {string} params.address - The address to fetch the ARNS names for
+   * @returns {Promise<AoArNSNameData[]>} The ARNS names associated with the address
+   */
+  async getArNSRecordsForAddress(
+    params: PaginationParams<AoArNSNameDataWithName> & {
+      antRegistryId?: string;
+      address: WalletAddress;
+    },
+  ): Promise<
+    PaginationResult<AoArNSNameDataWithName & { role: 'controller' | 'owner' }>
+  > {
+    const { antRegistryId = ANT_REGISTRY_ID, address } = params;
+    // create the ant registry
+    const antRegistry = ANTRegistry.init({
+      process: new AOProcess({
+        ao: this.process.ao,
+        processId: antRegistryId,
+      }),
+    });
+
+    // TODO: this could be memoized to avoid re-fetching the ACL for the same address
+    const { Controlled = [], Owned = [] } = await antRegistry.accessControlList(
+      {
+        address,
+      },
+    );
+
+    const allProcessIds = [...Controlled, ...Owned];
+
+    if (allProcessIds.length === 0) {
+      return {
+        items: [],
+        limit: 0,
+        totalItems: 0,
+        sortOrder: 'asc',
+        hasMore: false,
+      };
+    }
+
+    const namesForPage = await this.getArNSRecords({
+      ...params,
+      filters: {
+        // TODO: this filters cannot be larger than whatever AO allows for dry-runs, investigate and batch the requests if needed
+        processId: allProcessIds,
+      },
+    });
+
+    const namesWithRole: (AoArNSNameDataWithName & {
+      role: 'controller' | 'owner';
+    })[] = namesForPage.items.map((name) => ({
+      ...name,
+      role: Owned.includes(name.processId) ? 'owner' : 'controller',
+    }));
```

Comment:
I like the thought, but just want to mention this _might not be accurate_ if the ANT involved isn't sending correct state notices, or more likely, that the MU isn't cranking the notices.